src/api_sketch.jl

```diff
@@ -0,0 +1,210 @@
+
+# I think we should be able to do this too...there might be a more idiomatic way
+# to do what I'm showing below, perhaps via an "actual" holy trait type thing
+isatoms(::Type{AtomGraph}) = true
+isatoms(::Type{WeaveMol}) = true
+# add more for other types later on...
+isatoms(::Type{Any}) = false
+
+# magical encoding
+function (f<:Feature{Tn,Te})(atomsobj::T) where T
+    @assert isatoms(T) "Features can only be computed on atomic structures!"
+    f.encode_f(atomsobj)
+end
+
+# magical decoding
+(f<:Feature{Tn,Te})(encoded::Te) = f.decode_f(encoded)
+
+#=
+Feature of a single atom.
+
+(logspaced field not necessary because it can be embedded in encode_f and decode_f through 
+keywords to constructor (see below))
+=#
+struct AtomFeat{Tn,Te}<:Feature{Tn,Te}
+    name # symbol (for easy DataFrame indexing) or String (for easy other things)?
+    encode_f
+    decode_f
+    categorical::Bool
+    contextual::Bool # can get from elemental lookup table (false) or not (true)?
+    length::Int # length of encoded vector
+end
+
+#=
+we'll define a bunch of automatic stuff for building AtomFeats with built-in data
+that will essentially copy the current contents of atomfeat.jl, ideally with some
+additions enabling the user to augment the lookup table
+
+should also have convenience functions for building encode_f and decode_f via keywords for, e.g.:
+    choosing one-hot (and how many bins, etc.) vs. direct float encoding
+    logspaced vs. linear spaced binning for one-hot encoding
+    maybe other stuff too
+...
+=#
+
+# then basically copy a bunch of the weave stuff, plus potentially things like oxidation state etc. from pymatgen belong here too
+
+#=
+Feature of a pair of atoms. Currently only used in WeaveModel, but may eventually
+have a version of AtomGraph that allows edge features...
+=#
+struct PairFeat{Tn,Te}<:Feature{Tn,Te}
+    name
+    encode_f
+    decode_f
+    length::Int
+    # probably needs some other stuff...
+end
+# same idea here...copy from weave stuff...
+# a note: I think it's more elegant (and probably efficient?) if pair features assert
+# always arrays where the first two indices are atom indices and any further are indexing
+# into the feature if it is itself higher-dimensional. This is not how the weave featurizer
+# currently works...
+
+#=
+This is just an example of another sort of feature one might add, but it's not needed/used
+currently.
+=#
+struct StructureFeat{Tn,Te}<:Feature{Tn,Te}
+    name
+    encode_f
+    decode_f
+    # etc...
+end
+
+#= FEATURIZATION OBJECTS
+All such objects should define at least one list of <:Feature objects and a `combine`
+function that specifies how to put them together for eventual attachment to an atoms 
+object. 
+=#
+
+abstract type AbstractFeaturization end
+
+struct GraphNodeFeaturization
+    atom_feats::Vector{AtomFeat}
+    combine
+end
+
+struct WeaveFeaturization
+    atom_feats::Vector{AtomFeat}
+    pair_feats::Vector{PairFeat}
+    combine
+end
+
```

Comment:
`GraphNodeFeaturization` and `WeaveFeaturization` here are sub types of  `AbstractFeaturization`, right?